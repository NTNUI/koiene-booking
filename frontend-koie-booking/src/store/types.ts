// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
