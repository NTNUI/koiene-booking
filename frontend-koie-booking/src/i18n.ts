import Vue from 'vue';
import VueI18n, { LocaleMessages } from 'vue-i18n';
import enJson from './locales/en.json';

Vue.use(VueI18n);

function loadLocaleMessages(): LocaleMessages {
  // Testing currently only supports manually importing of json-files. English is the only language that gets loaded now
  let locales: any;
  const messages: LocaleMessages = {};
  if (process.env.NODE_ENV === 'test') {
    messages['en'] = enJson;
  } else {
    locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
    locales.keys().forEach((key: any) => {
      const matched = key.match(/([A-Za-z0-9-_]+)\./i);
      if (matched && matched.length > 1) {
        const locale = matched[1];
        messages[locale] = locales(key);
      }
    });
  }
  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages()
});
