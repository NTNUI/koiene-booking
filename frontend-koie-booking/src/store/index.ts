import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
