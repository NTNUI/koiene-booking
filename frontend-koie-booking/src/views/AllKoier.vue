<template>
  <ErrorCard v-if="error" />
  <LoadingSpinner v-else-if="isLoading" />
  <v-container v-else fill-height class="global-max-width pa-3">
    <v-row class="mx-auto">
      <v-col v-for="koie in koieData" :key="koie.name" class="mt-5" cols="12" sm="6" md="4">
        <v-card :title="koie.name" outlined class="global-card-outlined" @click="goToKoie">
          <v-img width="100%" height="200" :src="imgUrl(koie.name)"></v-img>
          <v-card-title>{{ koie.name }}</v-card-title>
          <v-card-subtitle>{{ koie.number_of_beds }} {{ $t('allKoier.beds') }} </v-card-subtitle>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import ErrorCard from '../components/ErrorCard.vue';
import LoadingSpinner from '../components/LoadingSpinner.vue';
import Vue from 'vue';
import { AllKoierData } from '../types/allKoier';

export default Vue.extend({
  name: 'AllKoier',
  components: {
    ErrorCard,
    LoadingSpinner,
  },
  data(): AllKoierData {
    return {
      koieTitle: '',
    };
  },
  computed: {
    error: function(): boolean {
      return this.$store.state.koie.error;
    },
    isLoading: function(): boolean {
      return this.$store.state.koie.isLoading;
    },
    koieData(): Array<any> {
      return this.$store.state.koie.allKoier;
    },
  },
  created: function() {
    this.$store.dispatch('koie/FETCH_ALL_KOIER');
    this.$store.dispatch('avalanche/DISABLE_BOOKING', false);
    if (this.$store.state.booking.beds === 0) {
      this.$store.dispatch('booking/SET_BEDS', 1);
      this.$store.dispatch('booking/SET_AVAILABLE_BEDS', 10);
    }
  },
  methods: {
    goToKoie(event: any) {
      this.koieTitle = event.currentTarget.title.toLowerCase();
      this.$router.push(`/koie/${this.koieTitle}`);
      this.getKoieData(this.koieTitle);
    },
    imgUrl(name: string): string {
      return '/koie_pictures/' + name.toLowerCase() + '1.jpg';
    },
    getKoieData(koieName: string) {
      this.$store.dispatch('koie/FETCH_DATA', koieName);
    },
  },
});
</script>

<style lang="scss"></style>
