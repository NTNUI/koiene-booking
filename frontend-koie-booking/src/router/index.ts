import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
