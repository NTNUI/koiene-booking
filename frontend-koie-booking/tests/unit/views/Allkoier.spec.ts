import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { storeConfig } from '@/store';
import i18n from '@/i18n';
import scssVars from '@/styles/variables.scss';
import flushPromises from 'flush-promises';
import mockAxios from 'jest-mock-axios';
import router from '@/router';
import { routes } from '@/router';
import { cloneDeep } from 'lodash';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue } from '@vue/test-utils';

// Components or views
import AllKoier from '../../../src/views/AllKoier.vue';

describe('View AllKoier.vue', () => {
  const router = new VueRouter({ routes, mode: 'abstract' });
  let wrapper: any;
  let localVue: any;
  let vuetify: any;
  let store: any;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(VueRouter);
    localVue.use(Vuetify);
    localVue.use(Vuex);
    // Resets the store between each test (necessary because of a data-saving check in the action that's used)
    // Use the config in Component_Koie.spec.ts if this setup is not necessary in your test
    store = new Vuex.Store(cloneDeep(storeConfig));
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    // cleaning up previous test's mock-api-calls
    mockAxios.reset();
  });

  it('Matches snapshot', async () => {
    wrapper = mount(AllKoier, {
      localVue,
      vuetify,
      router,
      i18n,
      store
    });

    const responseObj = {
      data: { koier: [{ name: 'flåkoia', number_of_beds: 10 }] }
    };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    expect(wrapper).toMatchSnapshot();
  });

  // This test is unnecessary, but serves as a good selector-example
  it('Displays card with text "flaakoia"', async () => {
    wrapper = mount(AllKoier, {
      localVue,
      vuetify,
      router,
      i18n,
      store
    });

    const responseObj = { data: { koier: [{ name: 'flåkoia', number_of_beds: 10 }] } };
    mockAxios.mockResponse(responseObj);

    const title = wrapper.find('[title="flåkoia"] > .v-card__title');

    await flushPromises();
    expect(title.text()).toBe('flåkoia');
  });

  it('Check routing to correct "Koie.vue"', async () => {
    wrapper = mount(AllKoier, {
      localVue,
      vuetify,
      router,
      i18n,
      store
    });
    const spy = jest.spyOn(wrapper.vm.$router, 'push');
    const responseObj = { data: { koier: [{ name: 'flåkoia', number_of_beds: 10 }] } };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    const title = wrapper.find('[title="flåkoia"]').trigger('click');

    expect(spy).toHaveBeenCalledWith('/koie/flåkoia');
  });
});
