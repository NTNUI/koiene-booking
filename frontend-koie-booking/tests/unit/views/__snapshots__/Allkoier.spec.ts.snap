// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View AllKoier.vue Matches snapshot 1`] = `
<div class="container global-max-width pa-3 fill-height">
  <div class="row mx-auto">
    <div class="mt-5 col-sm-6 col-md-4 col-12">
      <div tabindex="0" class="global-card-outlined v-card v-card--link v-sheet v-sheet--outlined theme--light" title="flåkoia">
        <div class="v-image v-responsive theme--light" style="height: 200px; width: 100%;">
          <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;" name="fade-transition" mode="in-out"></div>
          <div class="v-responsive__content"></div>
        </div>
        <div class="v-card__title">flåkoia</div>
        <div class="v-card__subtitle">10 beds </div>
      </div>
    </div>
  </div>
</div>
`;
